polaris-core/src/main/java/org/apache/polaris/core/catalog/PageToken.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.polaris.core.catalog;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * Represents a page token that can be used by operations like `listTables`. Clients that specify a
+ * `pageSize` (or a `pageToken`) may receive a `next-page-token` in the response, the content of
+ * which is a serialized PageToken.
+ *
+ * <p>By providing that in the next query's `pageToken`, the client can resume listing where they
+ * left off. If the client provides a `pageToken` or `pageSize` but `next-page-token` is null in the
+ * response, that means there is no more data to read.
+ */
+public class PageToken {
+
+  public final int offset;
+  public final int pageSize;
+
+  private static final String TOKEN_PREFIX = "polaris";
+  private static final int TOKEN_START = 0;
+  private static final int DEFAULT_PAGE_SIZE = 1000;
+
+  public static PageToken DONE = null;
+
+  public PageToken(int offset, int pageSize) {
+    this.offset = offset;
+    this.pageSize = pageSize;
+  }
+
+  /** Construct a PageToken from a plain limit */
+  public static PageToken fromLimit(int limit) {
+    return new PageToken(TOKEN_START, limit);
+  }
+
+  /** Construct a PageToken to read everything */
+  public static PageToken readEverything() {
+    return new PageToken(TOKEN_START, Integer.MAX_VALUE);
+  }
+
+  /** Deserialize a token string into a PageToken object */
+  public static PageToken fromString(String tokenString) {
+    if (tokenString == null) {
+      return PageToken.readEverything();
```

Comment:
What is the difference between an empty token and a non-existent token? Should they both behave the same by returning the DEFAULT_PAGE_SIZE?

Reply:
This was a little ambiguous to me upon reading the spec. [The spec says](https://github.com/apache/iceberg/blob/ab2c6f889d07eeee51a1f58605be248e9330d91b/open-api/rest-catalog-open-api.yaml#L1690):

> Clients may initiate the first paginated request by sending an empty query parameter `pageToken` to the server.

As the token is a string, I interpreted `empty query parameter` to map to an empty string.

I think `null` would be the same as the user not supplying the query parameter at all (?), and in such a case we want to use `PageToken.readEverything()`.

Reply:
I only concern is would we get into out of memory. I don't know in the internals that much. Assume there are 300K tables, do we see any concerns?

Reply:
The client is allowed to request a read without pagination. In that case, we could potentially OOM. But solving that is probably not in scope here; the goal of pagination as defined by the spec is to allow the client to _choose_ a smaller page size, not to enforce that it must choose a small page size.

Reply:
Thanks for clarifying. Sounds good. You can resolve my comments.